### PR TITLE
Add build script and automatic versioning

### PR DESCRIPTION
**Pull request by kevingelion**
This modification introduces build.sh - the build script that generates our version file and then either builds or builds AND uploads based on user input.

At the current state, this has **NOT** been confirmed to work on OS X yet.

cc: @jminardi @dreammaker @kdumontnu @riningerr

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -0,0 +1,143 @@
+#!/bin/bash
+# Correct Syntax: ./build.sh [COM port] [verify/build]
```
Comment (dreammaker) [inline build.sh]: Usually pipe is used to denote one of several options.  See `git help reset` for an example.

I think it's good to use `set -e` and `set -u` at the top of every shell script, which will exit the script if any command fails and if you try to use an undefined variable, respectively.

File: build.sh [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  eval "mv \"$ARDUINO_DEP/variants/rambo/\" \"$ARDUINO_DEP/variants/rambo_backup/\""
+fi
+# Even if a user doesn't have a rambo folder, they should have a boards.txt
+eval "mv \"$ARDUINO_DEP/boards.txt\" \"$ARDUINO_DEP/boards_backup.txt\""
+eval "cp \"$HERE/ArduinoAddons/Arduino_1.6.x/hardware/marlin/avr/boards.txt\" \"$ARDUINO_DEP/\""
+eval "cp -r \"$HERE/ArduinoAddons/Arduino_1.6.x/hardware/marlin/avr/variants/rambo/\" \"$ARDUINO_DEP/variants/rambo/\""
+
+# Create the build directory
+eval "mkdir \"$HERE/build/\""
+eval $ARDUINO_EXEC
```
Comment (dreammaker) [inline build.sh]: I don't think you need `eval` anywhere in this file.

Comment (kevingelion) [inline build.sh]: I know that `eval` has security risks, but am I using it dangerously? I'll try to remove all of them but if I remember correctly, Cygwin had some issues running a string as a command.

Comment (dreammaker) [inline build.sh]: I would be surprised since Cygwin runs bash by default.  You may just be running into edge cases more since you probably have spaces in file names when you're on Cygwin.

File: .gitignore [hunk 3]
```diff
@@ -1,6 +1,6 @@
 // Our automatic versioning scheme generates the following file
 // NEVER put it in the repository
-_Version.h
+Marlin/_Version.h
```
Comment (jminardi) [inline .gitignore]: Why was this change needed?

File: build.sh [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+HERE=$(pwd)
+VERSION='Voxel8 Marlin Build Script v1.0'
+
+# Begin build script
+if [ -z "$1" ]; then
+  echo "Please enter the port of the device or specify 'verify'"
+  exit 1
+fi
+if [ $1 = "upload" ]; then
+  echo "Syntax error. Please use --help for more info."
```
Comment (jminardi) [inline build.sh]: This happens when upload is specified without a port, correct? If so we should make the error message reflect that.

File: build.sh [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  echo $VERSION
+  echo "A star (*) represents a default value."
+  echo ""
+  echo "SYNTAX: build.sh [port [*upload | verify]]"
+  echo ""
+  echo "Type '--help name' to find out more about the argument 'name'."
+  exit 1
+elif [ $1 = "--help" ] && [ ! -z "$2" ]; then
+  if [ $2 = "port" ]; then
+    echo "This is the port of the device in which the sketch will be uploaded to. Can be emitted if just verifying."
```
Comment (jminardi) [inline build.sh]: typo: emitted -> omitted

Comment (kevingelion) [inline build.sh]: thanks, fixed

File: .gitignore (hunk 3 above)
Comment (kevingelion) [inline .gitignore]: I think i was under the impression that it wouldn't ignore _Version.h unless I specified the directory it was in. All fixed!

**Comment by jminardi** [comment]
This LGTM. Still waiting on @pizzyflavin's PR to merge, right?

**Comment by kevingelion** [comment]
Yep, and I'll continue working on getting the build script working on Linux.

**Comment by dreammaker** [comment]
Just thinking out loud... Since we're basically outputting code, would it make sense to use a template language like Jinja to output the `_Version.h` file?  In fact, most scripts this size I've found end up being significantly easier to write and maintain in a real programming language.  My personal choice for shell interaction would be ruby, but I'm guessing python would work just fine, especially since you're already using python for tests.

I'm guessing you decided against this so that there isn't a dependency on python to run the build, which makes it easier to support on multiple platforms.  Still thought it was worth considering.

**Comment by kevingelion** [comment]
I was originally going to write the build script in Python as there's a version generator script in `PlatformIOAddons/` already written. If I remember correctly, a few of us decided it was best to just do it all in bash. I can create one in Python if I need to.

**Comment by jminardi** [comment]
We considered python at first but we wanted to keep it more portable. I
agree this got bigger than I anticipated though.

On Wed, Oct 14, 2015 at 12:09 PM, Kevin Murphy <notifications@github.com>
wrote:

> I was originally going to write the build script in Python as there's a
> version generator script in PlatformIOAddons/ already written. If I
> remember correctly, a few of us decided it was best to just do it all in
> bash. I can create one in Python if I need to.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Voxel8/Marlin/pull/44#issuecomment-148101969>.
>



-- 
Jack Minardi
jack.minardi.org
jack@minardi.org

**Comment by jminardi** [comment]
This PR should not be held up with the Linux stuff, it should come in another PR
